### PR TITLE
Add fuzzing to CI

### PR DESCRIPTION
**Pull request by mborland**
As required by @anarthal conditional acceptance. CC: @ckormanyos

### CODE REVIEW

File: fuzzing/Jamfile [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+#        ;
+#}
+#else
+#{
+#    alias old-corpus ;
+#}
+#explicit old-corpus ;
+
+
+# local initial-corpus = [ glob-tree-ex ../test : *.json ] ;
```
Comment (anarthal) [inline fuzzing/Jamfile]: This is missing an initial corpus (like Url did some time ago). It's an important thing to provide it so that the fuzzer is adequately guided. You may find [this page](https://docs.cppalliance.org/contributor-guide/testing/fuzzing.html) useful, which I wrote some time ago when I implemented fuzzing for MySQL.

I suggest to extract the initial corpus from your unit tests.

File: fuzzing/Jamfile [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+
+local variants = from_chars_float from_chars_int ;
```
Comment (anarthal) [inline fuzzing/Jamfile]: You may also want to have a look at MySQL's Jamfile on fuzzing, which I think is shorter.

File: fuzzing/fuzz_from_chars_float.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <boost/charconv.hpp>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <exception>
+
+extern "C" int LLVMFuzzerTestOneInput(const std::uint8_t* data, std::size_t size)
+{
+    std::string s {reinterpret_cast<const char*>(data), size};
```
Comment (anarthal) [inline fuzzing/fuzz_from_chars_float.cpp]: Why do you need to copy this into a string? The faster the fuzzer, the better. And in real world, output is not null-terminated.

File: fuzzing/fuzz_from_chars_float.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    try
+    {
+        double val;
+        boost::charconv::from_chars(s.c_str(), s.c_str() + s.size(), val);
```
Comment (anarthal) [inline fuzzing/fuzz_from_chars_float.cpp]: You may want to add fuzzers for the different options you have for doubles. For instance, you can make the first byte of input generated by LLVM be the precision you apply to `from_chars`.

File: fuzzing/fuzz_from_chars_int.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#include <iostream>
+#include <exception>
+
+extern "C" int LLVMFuzzerTestOneInput(const std::uint8_t* data, std::size_t size)
+{
+    std::string s {reinterpret_cast<const char*>(data), size};
+
+    try
+    {
+        long val;
```
Comment (anarthal) [inline fuzzing/fuzz_from_chars_int.cpp]: You may also want to test with different int sizes.

File: fuzzing/Jamfile (hunk 1 above)
Comment (mborland) [reply fuzzing/Jamfile]: What order of magnitude of values need to be in the seed corpus? Like 10s or 1'000s?

Comment (anarthal) [reply fuzzing/Jamfile]: It's more about variety than quantity - the more different cases you cover, the better. AFAIK libfuzzer instruments your code to measure coverage for each sample. Samples that don't trigger any extra coverage don't add value. I don't know exactly what metric they use to measure coverage. They recommend to add both well-formed and malformed samples. 

TL;DR: probably >10 and <1000

**Comment by mborland** [comment]
@anarthal Let me know if there's anything else you wanted added to this; otherwise I think it's good to merge.

**Comment by anarthal** [comment]
Looks good to me.

**Comment by pdimov** [comment]
I don't see any actual fuzzing being performed in https://github.com/cppalliance/charconv/actions/runs/7651955730. Looks like the executables are just compiled and linked, but not run.

**Comment by mborland** [comment]
> I don't see any actual fuzzing being performed in https://github.com/cppalliance/charconv/actions/runs/7651955730. Looks like the executables are just compiled and linked, but not run.

As usual you are correct and it has been fixed. Fuzzer found two issues which have both been addressed.
